cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java

```diff
@@ -0,0 +1,308 @@
+  private final HBaseQueueClientFactory queueClientFactory;
+
+  @Inject
+  public HBaseQueueDebugger(HBaseQueueAdmin queueAdmin,
+                            HBaseQueueClientFactory queueClientFactory,
+                            TransactionServiceClient txClient,
+                            ZKClientService zkClientService) {
+    this.queueAdmin = queueAdmin;
+    this.queueClientFactory = queueClientFactory;
+    this.txClient = txClient;
+    this.zkClientService = zkClientService;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    zkClientService.startAndWait();
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    zkClientService.stopAndWait();
+  }
+
+  /**
+   * Only works for {@link co.cask.cdap.data2.transaction.queue.hbase.ShardedHBaseQueueStrategy}.
+   */
+  private void scanQueue(URI queueNameURI) throws Exception {
+    QueueName queueName = QueueName.from(queueNameURI);
+    HBaseConsumerStateStore stateStore = queueAdmin.getConsumerStateStore(queueName);
+
+    Transaction tx = txClient.startLong();
+    stateStore.startTx(tx);
+
+    Multimap<Long, QueueBarrier> barriers = stateStore.getAllBarriers();
+    LOG.info("Got {} barriers", barriers.size());
+
+    QueueStatistics stats = new QueueStatistics();
+    int currentSection = 1;
+
+    for (Long groupId : barriers.keySet()) {
+      LOG.info("Scanning barriers for group {}", groupId);
+      Collection<QueueBarrier> groupBarriers = barriers.get(groupId);
+      Iterator<QueueBarrier> barrierIterator = groupBarriers.iterator();
+
+      QueueBarrier previous;
+      QueueBarrier current = null;
+      while (barrierIterator.hasNext()) {
+        previous = current;
+        current = barrierIterator.next();
+
+        if (previous != null) {
+          // after first iteration
+          LOG.info("Scanning section {}/{}...", currentSection, barriers.size());
+          QueueStatistics sectionStats = scanQueue(tx, queueName, previous, current);
+          LOG.info("Section results: {}", sectionStats.getReport());
+
+          stats = QueueStatistics.of(stats, sectionStats);
+          currentSection++;
+        }
+      }
+      // handle last barrier
+      LOG.info("Scanning section {}/{}...", currentSection, barriers.size());
+      stats = QueueStatistics.of(stats, scanQueue(tx, queueName, current, null));
+      LOG.info("Scanning complete", currentSection, barriers.size());
+    }
+
+    LOG.info("Total results: {}", stats.getReport());
+    stateStore.commitTx();
+  }
+
+  private QueueStatistics scanQueue(Transaction tx, QueueName queueName, QueueBarrier start,
+                                    @Nullable QueueBarrier end) throws IOException {
+
+    byte[] queueRowPrefix = QueueEntryRow.getQueueRowPrefix(queueName);
+
+    ConsumerGroupConfig consumerConfig = start.getGroupConfig();
+    LOG.info("Got consumer group config: {}", consumerConfig);
+
+    HBaseQueueAdmin admin = queueClientFactory.ensureTableExists(queueName);
```

Comment:
If the queue table doesn't exists, this tool shouldn't creating it. It should just return (as if there is no queue entry).